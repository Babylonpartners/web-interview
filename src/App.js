import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import { formatRelative } from 'date-fns'
import { enGB } from 'date-fns/locale'

import {
  getUsers,
  getAvailableSlots,
  selectConsultantType,
  selectTimeSlot,
  selectAppointmentType,
  setAppointmentNotes,
  postAppointment,
} from './actions/newAppointment'

import './App.scss'

import AppHeader from './components/AppHeader/AppHeader'
import Separator from './components/Separator/Separator'

import UserHeader from './components/UserHeader/UserHeader'
import LabeledField from './components/LabeledField/LabeledField'
import StethoscopeIcon from './components/Icons/StethoscopeIcon'
import TimeIcon from './components/Icons/TimeIcon'
import CameraIcon from './components/Icons/CameraIcon'
import DocumentIcon from './components/Icons/DocumentIcon'
import PhotoIcon from './components/Icons/PhotoIcon'
import Button from './components/Button/Button'
import TextArea from './components/TextArea/TextArea'
import RadioGroup from './components/RadioGroup/RadioGroup'

class App extends Component {
  constructor(props) {
    super(props)

    this.handleConsultantTypeChange = this.handleConsultantTypeChange.bind(this)
    this.handleAppointmentTypeChange = this.handleAppointmentTypeChange.bind(
      this
    )
    this.handleTimeSlotChange = this.handleTimeSlotChange.bind(this)
    this.handleNotesChange = this.handleNotesChange.bind(this)
    this.handleBookAppointment = this.handleBookAppointment.bind(this)
  }

  componentDidMount() {
    const { getUsers, getAvailableSlots } = this.props
    getUsers()
    getAvailableSlots()
  }

  handleConsultantTypeChange(e) {
    this.props.selectConsultantType(e.target.value)
  }

  handleAppointmentTypeChange(e) {
    this.props.selectAppointmentType(e.target.value)
  }

  handleTimeSlotChange(e) {
    const selected = this.props.availableSlots.find(
      slot => slot.id.toString() === e.target.value
    )
    debugger
    this.props.selectTimeSlot(selected)
  }

  handleNotesChange(e) {
    this.props.setAppointmentNotes(e.target.value)
  }

  handleBookAppointment(e) {
    const {
      postAppointment,
      selectedUser,
      selectedConsultantType,
      selectedTimeSlot,
      appointmentNotes,
    } = this.props

    postAppointment({
      userId: selectedUser.id,
      dateTime: selectedTimeSlot.time,
      notes: appointmentNotes,
      type: selectedConsultantType,
    })
  }

  parseConsultantType(slots = []) {
    const consultantTypes = []
    slots.forEach(slot => {
      slot.consultantType.forEach(type => {
        if (!consultantTypes.includes(type)) {
          consultantTypes.push(type)
        }
      })
    })
    return consultantTypes.map(type => [
      type.charAt(0).toUpperCase() + type.slice(1),
      type,
    ])
  }

  parseTimeSlots(slots = []) {
    return slots
      .filter(slot =>
        slot.consultantType.includes(this.props.selectedConsultantType)
      )
      .map(slot => {
        const dateFmt = formatRelative(new Date(slot.time), new Date(), {
          locale: enGB,
        })
        return [
          dateFmt.charAt(0).toUpperCase() + dateFmt.slice(1),
          slot.id.toString(),
        ]
      })
  }

  render() {
    const {
      usersLoading,
      selectedUser,
      availableSlotsLoading,
      availableSlots,
      selectedConsultantType,
      selectedTimeSlot,
      selectedAppointmentType,
      appointmentNotes,
    } = this.props

    const consultantTypes = this.parseConsultantType(availableSlots)
    console.log('consultantTypes', consultantTypes)

    const appointmentTypes = [['Video', 'video'], ['Audio', 'audio']]

    const timeSlots = this.parseTimeSlots(availableSlots)

    return (
      <div className="app">
        <AppHeader />
        <main>
          <header>New Appointment</header>
          {usersLoading && <div>LOADING...</div>}
          {!usersLoading && (
            <UserHeader
              name={`${selectedUser.firstName} ${selectedUser.lastName}`}
              imgSrc={selectedUser.avatar}
            />
          )}

          <Separator />

          <LabeledField icon={<StethoscopeIcon />} label="Consultant Type">
            <RadioGroup
              legend="Consultant Type"
              inputName="consultantType"
              options={consultantTypes}
              selectedValue={selectedConsultantType}
              onChange={this.handleConsultantTypeChange}
            />
          </LabeledField>

          <LabeledField icon={<TimeIcon />} label="Date & Time">
            {!availableSlotsLoading && (
              <RadioGroup
                legend="Date & Time"
                inputName="appointmentSlot"
                options={timeSlots}
                selectedValue={`${selectedTimeSlot.id}`}
                onChange={this.handleTimeSlotChange}
              />
            )}
          </LabeledField>

          <LabeledField icon={<CameraIcon />} label="Appointment Type">
            <RadioGroup
              legend="Appointment Type"
              inputName="appointmentType"
              options={appointmentTypes}
              selectedValue={selectedAppointmentType}
              onChange={this.handleAppointmentTypeChange}
            />
          </LabeledField>

          <LabeledField icon={<DocumentIcon />} label="Notes">
            <TextArea
              value={appointmentNotes}
              onChange={this.handleNotesChange}
            />
          </LabeledField>

          <LabeledField icon={<PhotoIcon />} label="Attach a photo">
            <Button invert>+</Button>
          </LabeledField>

          <Separator />

          <Button fullwidth onClick={this.handleBookAppointment}>
            Book
          </Button>
        </main>
      </div>
    )
  }
}

const timeSlotType = PropTypes.shape({
  id: PropTypes.number,
  consultantType: PropTypes.arrayOf(PropTypes.string),
  appointmentType: PropTypes.arrayOf(PropTypes.string),
  time: PropTypes.string,
})

App.propTypes = {
  usersLoading: PropTypes.bool,
  selectedUser: PropTypes.shape({
    id: PropTypes.number,
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    avatar: PropTypes.string,
  }),
  availableSlots: PropTypes.arrayOf(timeSlotType),
  availableSlotsLoading: PropTypes.bool,
  selectedConsultantType: PropTypes.string,
  selectedTimeSlot: timeSlotType,
  selectedAppointmentType: PropTypes.string,
  appointmentNotes: PropTypes.string,

  getUsers: PropTypes.func,
  getAvailableSlots: PropTypes.func,
  selectConsultantType: PropTypes.func,
  selectTimeSlot: PropTypes.func,
  selectAppointmentType: PropTypes.func,
  setAppointmentNotes: PropTypes.func,
  postAppointment: PropTypes.func,
}

const mapStateToProps = state => ({
  usersLoading: state.newAppointment.usersLoading,
  selectedUser: state.newAppointment.selectedUser,
  availableSlots: state.newAppointment.availableSlots,
  availableSlotsLoading: state.newAppointment.availableSlotsLoading,
  selectedConsultantType: state.newAppointment.selectedConsultantType,
  selectedTimeSlot: state.newAppointment.selectedTimeSlot,
  selectedAppointmentType: state.newAppointment.selectedAppointmentType,
  appointmentNotes: state.newAppointment.appointmentNotes,
})

const mapDispatchToProps = dispatch => ({
  getUsers: () => dispatch(getUsers()),
  getAvailableSlots: () => dispatch(getAvailableSlots()),
  selectConsultantType: consultantType =>
    dispatch(selectConsultantType(consultantType)),
  selectTimeSlot: timeSlot => dispatch(selectTimeSlot(timeSlot)),
  selectAppointmentType: appointmentType =>
    dispatch(selectAppointmentType(appointmentType)),
  setAppointmentNotes: appointmentNotes =>
    dispatch(setAppointmentNotes(appointmentNotes)),
  postAppointment: appointment => dispatch(postAppointment(appointment)),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)
