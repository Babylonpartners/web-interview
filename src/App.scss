header {
  img {
    padding: 12px;
  }
  border-bottom: solid 1px #dddddd;
}

.booking-container {
  padding-left: 12px;
}
