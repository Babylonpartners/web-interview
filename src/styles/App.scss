@import './constants.scss';

// shared styles


.button {
  box-sizing: border-box;
  padding: 12px;
  background-color: $positive-light;
  color: $white;
  width: calc(25% - 12px);
  height: 42px;
  margin-right: 12px;
  display: inline-block;
  text-align: center;
  border-radius: 25px;
  box-shadow: none;
  outline: none;
}


.mobile-button {
  box-sizing: border-box;
  padding: 12px;
  background-color: $positive-light;
  color: $white;
  width: calc(25% - 12px);
  height: 42px;
  margin-right: 12px;
  display: block;
  text-align: center;
  border-radius: 25px;
  width: 100%;
  box-shadow: none;
  outline: none;
}

.white-button {
  box-sizing: border-box;
  padding: 12px;
  background-color: $white;
  border-color: $light-grey;
  color: $light-grey-font;
  width: calc(25% - 12px);
  height: 42px;
  margin-right: 12px;
  display: inline-block;
  text-align: center;
  border-radius: 25px;
  box-shadow: none;
  outline: none;
}

.mobile-white-button {
  box-sizing: border-box;
  padding: 12px;
  background-color: $white;
  border-color: $light-grey;
  color: $light-grey-font;
  width: calc(25% - 12px);
  height: 42px;
  margin-right: 12px;
  display: block;
  text-align: center;
  border-radius: 25px;
  width: 100%;
  box-shadow: none;
  outline: none;
}


h2 {
  &.h6 {
    font-family: "'Roboto', sans-serif";
    font-size: 24px;
  }
}

// components

.app {
  margin-left: 80px;
  margin-right: 80px;
}

.app-logo {
  display: block;
  margin-left: auto;
  margin-right: auto;
  width: 40%;
  height: 50px;
  width: 200px;
  margin-top: 15px;
}

.app-nav-bar {
  padding: $gutter-size--large;

  img {
    max-height: 33px;
  }

  border-bottom: $border-standard;
}



.new-appointment-header {
  font-family: "'Roboto', sans-serif";
  font-size: 50px;
  color: $header-grey-font;
}

.mobile-new-appointment-header {
  font-family: "'Roboto', sans-serif";
  font-size: 30px;
  color: $header-grey-font;
}

.user-panel {
  margin-top: 60px;
}

.user-panel img {
  float: left;
  width: 60px;
  height: 60px;
  margin-right: 15px;
}

.user-name {
  margin-top: 20px;
  vertical-align: middle;
}

.user-panel {
  margin-bottom: 60px;
}

.consultant-type-field {
  padding: $gutter-size--large;
  border-bottom: $border-standard;
}

.date-and-time-field {
  padding: $gutter-size--large;
  border-bottom: $border-standard;
}

.appointment-type-field {
  padding: $gutter-size--large;
  border-bottom: $border-standard;
}

.notes {
  padding-left: 15px;
  padding-top: 15px;
  margin-top: 40px;
  margin-bottom: 40px;
  height: 180px;
  display: block;
  width: 100%;
  border-radius: 15px;
  border-color: grey;
  box-shadow: none;
  background-color: transparent;
  resize: none;
  outline: none;
}

.submit-button {
  box-sizing: border-box;
  padding: 12px;
  background-color: $positive-light;
  color: $white;
  width: calc(25% - 12px);
  height: 42px;
  margin-right: 12px;
  text-align: center;
  border-radius: 15px;
  margin-top: 30px;
  display: block;
  width: 100%;
  box-shadow: none;
  outline: none;
}


