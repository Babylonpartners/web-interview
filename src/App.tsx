import React, { Component } from 'react'
import { BrowserRouter as Router, Route } from 'react-router-dom'
import logo from './logo.png'
import { API_ENDPOINT } from './config'

import './App.scss'

import Home from './pages/Home'
import Appointments from './pages/Appointments'
import BookAppointment from './pages/BookAppointment'

class App extends Component {
  public componentDidMount() {
    fetch(`${API_ENDPOINT}/users/1`)
      .then(res => res.json())
      .then(user => {
        // expect the logged in user object here
      })
      .catch(() => {
        // Use an Error boundary here
      })
  }

  public render() {
    return (
      <Router>
        <div className="app">
          <header className="app-header">
            <img src={logo} className="app-logo" alt="logo" />
          </header>
          <>
            <Route exact path="/" component={Home} />
            <Route exact path="/appointments" component={Appointments} />
            <Route exact path="/book" component={BookAppointment} />
          </>
        </div>
      </Router>
    )
  }
}

export default App
