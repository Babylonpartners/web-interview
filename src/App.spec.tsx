import React from 'react'
import renderer from 'react-test-renderer'
import ReactSixteenAdapter from 'enzyme-adapter-react-16'
import * as enzyme from 'enzyme'

enzyme.configure({ adapter: new ReactSixteenAdapter() })

import App from './App'

describe('App', () => {
  const element = <App />

  let component

  beforeEach(() => {
    const wrapper = enzyme.shallow(element)

    component = wrapper.instance()

    component.modelService = {
      get: jest.fn(key => {
        if (key === 'availableSlots') {
          return Promise.resolve([
            {
              id: 1,
              consultantType: ['GP'],
              appointmentType: ['Audio', 'Video'],
              time: '2019-11-27T10:11:00.000Z',
            },
            {
              id: 2,
              consultantType: ['Specialist', 'GP'],
              appointmentType: ['Audio', 'Video'],
              time: '2019-12-01T14:16:30.000Z',
            },
          ])
        }

        if (key === 'users') {
          return Promise.resolve({
            id: 1,
            firstName: 'Jane',
            lastName: 'Doe',
            avatar: '',
          })
        }
      }),
      post: jest.fn(),
    }
  })

  it('should match snapshot', () => {
    const snapshotComponent = renderer.create(element)
    expect(snapshotComponent.toJSON()).toMatchSnapshot()
  })

  describe('Component Did Mount', () => {
    it('should update the state with the available slots from the API', () => {
      component.componentDidMount()
    })
  })
})
