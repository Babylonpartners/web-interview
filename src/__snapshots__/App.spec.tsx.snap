// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="app"
>
  <div
    className="top-bar"
  >
    <button
      className="top-bar__burger"
    >
      <span
        className="icon icon--burger"
        title="burger"
      />
    </button>
    <div
      className="top-bar__logo"
    >
      <img
        alt="Babylon is awesome"
        src="logo.png"
      />
    </div>
  </div>
  <h1
    className="heading heading--large"
  >
    New Appointment
  </h1>
  <div
    className="form"
  >
    <div
      className="group"
    >
      <div
        className="user"
      >
        <img
          alt=""
          className="user__avatar"
          src=""
        />
        <span
          className="user__name"
        >
           
        </span>
      </div>
      <div
        className="divider"
      />
      <div
        className="form-field form-field--radio"
      >
        <div
          className="form-field__label"
        >
          <span
            className="icon icon--stethoscope"
            title="stethoscope"
          />
          <span>
            Consultant Type
          </span>
        </div>
        <div
          className="form-field__input"
        />
      </div>
    </div>
    <div
      className="group"
    >
      <div
        className="form-field form-field--radio"
      >
        <div
          className="form-field__label"
        >
          <span
            className="icon icon--clock"
            title="clock"
          />
          <span>
            Date & Time
          </span>
        </div>
        <div
          className="form-field__input"
        />
      </div>
      <div
        className="form-field form-field--radio"
      >
        <div
          className="form-field__label"
        >
          <span
            className="icon icon--radio"
            title="radio"
          />
          <span>
            Appointment Type
          </span>
        </div>
        <div
          className="form-field__input"
        />
      </div>
      <div
        className="form-field form-field--textarea"
      >
        <div
          className="form-field__label"
        >
          <span
            className="icon icon--video"
            title="video"
          />
          <span>
            Notes
          </span>
        </div>
        <textarea
          className="form-field__input"
          onChange={[Function]}
          placeholder="Describe your symptoms"
        />
      </div>
    </div>
  </div>
  <div
    className="bottom-bar"
  >
    <div
      className="divider"
    />
    <button
      className="button button--submit"
      onClick={[Function]}
    >
      Book
    </button>
  </div>
</div>
`;
